cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeEngineCredentials.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.runtime.spi.provisioner.dataproc;
+
+import com.google.api.client.util.GenericData;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.gson.Gson;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpRequests;
+import io.cdap.common.http.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Fecth and refresh token
+ */
+public class ComputeEngineCredentials {
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeEngineCredentials.class);
+  private static final Gson GSON = new Gson();
+
+  /**
+   * Time (in millisecond) to refresh the credentials before it expires.
+   */
+  private static final long REFRESH_BEFORE_EXPIRATION_TIME = 1000 * 60 * 10;
+  private static final RetryStrategy RETRY_STRATEGY =
+    RetryStrategies.limit(10, RetryStrategies.exponentialDelay(5, 10, TimeUnit.SECONDS));
+  private static GoogleCredentials cachedGoogleCredentials = null;
+
+  public static GoogleCredentials getComputeEngineCredentials(String endPoint) throws IOException {
+    if (cachedGoogleCredentials != null && System.currentTimeMillis() <
+      cachedGoogleCredentials.getAccessToken().getExpirationTime().getTime() - REFRESH_BEFORE_EXPIRATION_TIME) {
+      // Cached GoogleCredentials is still valid.
+      return cachedGoogleCredentials;
+    }
+
+    if (endPoint != null) {
+      try {
+        cachedGoogleCredentials = Retries.callWithRetries((Retries.Callable<GoogleCredentials, Exception>)
+                                                            () -> getDataprocCredentials(endPoint), RETRY_STRATEGY,
+                                                          Retries.ALWAYS_TRUE);
+        return cachedGoogleCredentials;
+      } catch (Exception e) {
+        LOG.error("Failed to fetch compute engine credentials", e);
+        return null;
+      }
+    }
+
+    try {
```

Comment:
So we only reach this try-catch if we're not using the new pod-isolation logic right? Maybe add a comment to make that more clear

Reply:
Done!